mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,110 +19,446 @@
+
+            # reset Godeps.json?
+            if [ -n "$(godep-changes ${f_mainline_commit})" ]; then
+                reset-godeps ${f_mainline_commit}^
+                dst_needs_godeps_update=true
+            fi
+
+            # finally cherry-pick
+            if ! git-with-user cherry-pick --keep-redundant-commits ${f_mainline_commit} >/dev/null; then
+                echo
+                show-working-dir-status
+                return 1
+            fi
+
+            # if there is no pending merge commit, update Godeps.json because this could be a target of tag
+            if [ -z "${k_pending_merge_commit}" ]; then
+                fix-godeps "${deps}" "${is_library}" ${dst_needs_godeps_update}
+                dst_needs_godeps_update=false
+                dst_merge_point_commit=$(git rev-parse HEAD)
+            fi
+        else
+            local f_mainline_merge=${f_mainline_commit}
+
+            # find **latest** (in the sense of least distance to ${f_mainline_merge}) common ancestor of both parents of ${f_mainline_merge}.
+            # If the PR had no merge commits, this is the actual fork point. If there have been merges, we will get the
+            # latest of those merges. This will potentially drop commit history, but no actual changes.
+            local f_latest_branch_point_commit=$(git merge-base --octopus ${f_mainline_merge}^1 ${f_mainline_merge}^2)
+            if [ -z "${f_latest_branch_point_commit}" ]; then
+                echo "Not branch point found for PR merged through ${f_mainline_merge}. Odd."
+                return 1
+            fi
+
+            # start cherry-picking with latest merge, squash everything before
+            local f_first_pick_base=${f_latest_branch_point_commit}
+            local f_latest_merge_commit=$(git log --merges --format='%H' --ancestry-path -1 ${f_latest_branch_point_commit}..${f_mainline_merge}^2)
+            if [ -n "${f_latest_merge_commit}" ]; then
+                echo "Cherry-picking squashed k8s.io/kubernetes commits $(kube-commit ${f_latest_branch_point_commit})..$(kube-commit ${f_latest_merge_commit}) because the last one is a merge: $(git show --format="%s" -q ${f_latest_merge_commit})"
+
+                # reset Godeps.json?
+                if [ -n "$(godep-changes ${f_latest_branch_point_commit} ${f_latest_merge_commit})" ]; then
+                    reset-godeps ${f_latest_branch_point_commit}
+                    dst_needs_godeps_update=true
+                fi
+
+                if ! git diff ${f_latest_branch_point_commit} ${f_latest_merge_commit} | git apply --index; then
+                    echo
+                    show-working-dir-status
+                    return 1
+                fi
+                git-with-user commit -q -m "sync: squashed up to merge $(kube-commit ${f_latest_merge_commit}) in $(kube-commit ${f_mainline_merge})"
```

Comment:
What do you mean by `in $(kube-commit ${f_mainline_merge}`?

Reply:
that's the kube commit of the github PR merge commit we are working on. Not sure how to express that better, maybe we can also drop this as it is confusing.